Add custom item system (nnextcitem.cpp, nnextcitem.h)

NoOneBlood opened the pull request:
Replace IniFile class (inifile.cpp, inifile.h)

NoOneBlood commented:
Any other notes or i can merge it?

Hendricks266 commented on source/blood/src/nnextcitem.cpp:
```diff
@@ -0,0 +1,1645 @@
+
+enum enum_ITEM_GROUP
+{
+kItemGroupItem             = 0,
+kItemGroupWeapon,
+kItemGroupAmmo,
+kItemGroupArmor,
+kItemGroupHealth,
+kItemGroupPowerup,
+kItemGroupKey,
+kItemGroupMax,
```
Can you fix the indentation of the enums in this file?

Hendricks266 commented on source/blood/src/nnextcitem.h:
```diff
@@ -0,0 +1,158 @@
+struct ITEM
+{
+    char *name, *message;
+	uint8_t  group, type;
```
Formatting

Hendricks266 commented on source/blood/src/nnextcitem.h:
```diff
@@ -0,0 +1,158 @@
+        static char  SetupActionLimits(ITEM::ACTION* pAct, char extLimits);
+        static char  SetupAction(const char* str, int nOperator, int nAction);
+        static char  SetupFlags(const char* str);
+		static char  SetupRespawn(const char* str);
```
Formatting

Hendricks266 commented on source/blood/src/nnextcitem.h:
```diff
@@ -0,0 +1,158 @@
+        static char showLog;
+//-------------------------------------------------------------------------------
+        static int  Setup(IniFile* pFile);
+		static void ClearItem(ITEM* pItem);
```
Formatting

Hendricks266 reviewed (commented):
Can you tell me a little bit about the changes to the INI parser? Did you change or expand the grammar and syntax of the file, beyond the addition of new fields?

NoOneBlood commented:
> Can you tell me a little bit about the changes to the INI parser? Did you change or expand the grammar and syntax of the file, beyond the addition of new fields?

The only real reason why i replaced it, is because the original one is pretty buggy. It does not clear spaces on the end of the string and ResReadLine function can go out of bounds if it doesn't contain null byte or end of srting characters in the end of file (when you load INI as raw data using resource file system for example).

The other reasons is that i added GetNextString and GetNextSection members for faster lookup.

Final reason is that it easier for me to update it later.

NoOneBlood commented on source/blood/src/actor.cpp:
```diff
@@ -2468,6 +2468,23 @@ void actInit(bool bSaveLoad) {
         nnExtInitModernStuff(bSaveLoad);
     }
+
+    if (VanillaMode())
+    {
+        if (gUserItemsInitialized)
+        {
+            userItemsUninit();
+            gUserItemsInitialized = 0;
+        }
+    }
```
It's a good question i don't know how to answer. Paranoid mode? There is always a chance that custom items will be loaded when user don't want it. I can remove vanilla mode check which makes it even better.

NoOneBlood commented on source/blood/src/inifile.cpp:
```diff
@@ -29,438 +29,552 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
-IniFile::IniFile(const char *fileName)
+
+void strTrim(char* str, char side = 0x03)
 {
-    head.next = &head;
+	int l = strlen(str);
+	int c;
+	
+	if (side & 0x01)
+	{
+		c = 0;
+		while(str[c] && isspace(str[c])) c++;
+		if (c)
+			memcpy(str, &str[c], l - c + 1), l-=c;
+	}
+	
+	
+	if (side & 0x02)
```
Not really. It's unrelated.

Hendricks266 commented on source/blood/src/nnextcitem.cpp:
```diff
@@ -482,21 +482,14 @@ int userItemGetRespawnTime(spritetype* pSpr)
 
-        if (gGameOptions.nItemSettings == 2)
-            return 1;
+        return (gGameOptions.nItemSettings == 2);
```
Good call on simplifying these return statements. It would be even better to remove the parentheses, because if someone made a typo and changed the `==` to `=`, having the statement wrapped in parentheses would silence compiler warnings about it.

Hendricks266 commented on source/blood/src/actor.cpp:
```diff
@@ -2621,7 +2611,6 @@ void actInit(bool bSaveLoad) {
 
         #ifdef NOONE_EXTENSIONS
-        if (!VanillaMode())
             userItemsInitSprites();
```
Could you fix the indentation here, now that the `if` statement is removed?

Hendricks266 commented on source/blood/src/inifile.h:
```diff
@@ -90,11 +90,11 @@ class IniFile
 		char  PutKeyInt(const char* section, const char* hiWord, const int nVal);
 		char  PutKeyHex(const char* section, const char* hiWord, const int nVal);
 		//---------------------------------------------------------------
```
Could you convert tabs to 4 spaces throughout this file?

Hendricks266 commented on source/blood/src/actor.cpp:
```diff
@@ -2915,7 +2904,7 @@ spritetype *actDropObject(spritetype *pSprite, int nType) {
 #ifdef NOONE_EXTENSIONS
-    if (!VanillaMode() && IsUserItem(nType))
+    if (/*!VanillaMode() &&*/ IsUserItem(nType))
     {
```
Could you remove the three instances of `/*!VanillaMode() &&*/ ` comments? I don't see a reason to keep them.

Hendricks266 commented on source/blood/src/actor.cpp:
```diff
@@ -6303,7 +6292,7 @@ int actGetRespawnTime(spritetype *pSprite) {
 
 #ifdef NOONE_EXTENSIONS
-    if (!VanillaMode() && IsUserItemSprite(pSprite))
+    if (/*!VanillaMode() &&*/ IsUserItemSprite(pSprite))
```
See above

Hendricks266 commented on source/blood/src/player.cpp:
```diff
@@ -1379,7 +1379,7 @@ void PickUp(PLAYER *pPlayer, spritetype *pSprite)
     }
 
-    if (!VanillaMode() && IsUserItem(nType))
+    if (/*!VanillaMode() &&*/ IsUserItem(nType))
```
See above

Hendricks266 reviewed (commented):
Thanks for your changes so far. This is taking shape.
